sdk/core/Azure.Core/src/Http/ConditionalRequestOptions.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.Core.Http
+{
+    /// <summary>
+    /// Specifies HTTP options for conditional requests.
+    /// </summary>
+    public class ConditionalRequestOptions : IEquatable<ConditionalRequestOptions>
+    {
+        private const string IfMatchHeader = "If-Match";
+        private const string IfNoneMatchHeader = "If-None-Match";
+
+        /// <summary>
+        /// Optionally limit requests to resources that have a matching ETag.
+        /// </summary>
+        public ETag? IfMatch { get; set; }
+
+        /// <summary>
+        /// Optionally limit requests to resources that do not match the ETag.
+        /// </summary>
+        public ETag? IfNoneMatch { get; set; }
+
+        /// <summary>
+        /// Set preconditions that indicate to apply an operation only if the
+        /// target resource has changed.
+        /// </summary>
+        /// <param name="etag">The version of the resource to compare to.</param>
+        public virtual void SetIfModifiedCondition(ETag etag)
+        {
+            IfNoneMatch = etag;
+        }
+
+        /// <summary>
+        /// Set preconditions that indicate to apply an operation only if the
+        /// target resource has not changed.
+        /// </summary>
+        public virtual void SetIfUnmodifiedCondition(ETag etag)
+        {
+            IfMatch = etag;
+        }
+
+        /// <summary>
+        /// Set preconditions that indicate to apply an operation only if the
+        /// target resource is present.
+        /// </summary>
+        public virtual void SetIfPresentCondition()
+        {
+            IfMatch = new ETag("*");
+        }
+
+        /// <summary>
+        /// Set preconditions that indicate to apply an operation only if the
+        /// target resource is not present.
+        /// </summary>
+        public virtual void SetIfAbsentCondition()
+        {
+            IfNoneMatch = new ETag("*");
+        }
+
+        public virtual void ApplyHeaders(Request request)
```

Comment:
Should this be internal?  And overridden by `TemporalRequestOptions`?

Reply:
Can't be internal, because it's in Azure.Core and needs to be called from the client library ... unless you see something I don't?

Reply:
It's more of a question of "is this a feature customers should use" or "is this to help us implement clients"?  We typically make the latter APIs internal helpers and share their source via Includes.

Reply:
@pakrym How do you feel about moving this into Shared?

Reply:
I'm for it